src/Language/PureScript/CoreFn/Laziness.hs

```diff
@@ -0,0 +1,504 @@
+-- the entire tree aborts.
+--
+-- One might ask, why interleave the abort monad with the tree at all if we're
+-- just going to flatten it out at the end? The point is to flatten it out at
+-- the end, but *not* during the generation of the tree. Attempting to flatten
+-- the tree as we generate it can result in an infinite loop, because a subtree
+-- needs to be exhaustively searched for abort conditions before it can be used
+-- in another tree. With this approach, we can use lazy trees as building
+-- blocks and, as long as they get rewritten to be finite or have aborts before
+-- they're flattened, the analysis still terminates.
+
+-- |
+-- Given a maximum index and a function that returns a map of edges to next
+-- indices, returns an array for each index up to maxIndex of maps from the
+-- indices reachable from the current index, to the maximum force applied to
+-- those indices.
+searchReachable
+  :: forall m force
+   . (Alternative m, Monad m, Enum force, Ord force)
+  => Int
+  -> ((Int, force) -> m (IM.MonoidalIntMap (Max force)))
+  -> A.Array Int (m (IM.MonoidalIntMap (Max force)))
+searchReachable maxIdx lookupEdges = mrtFlatten . unsafeHead <$> mem
+  where
+  -- This is a finite array of infinite lists, used to memoize all the search
+  -- trees. `unsafeHead` is used above to pull the first tree out of each list
+  -- in the array--the one corresponding to zero force, which is what's needed
+  -- to initialize the corresponding identifier. (`unsafeHead` is safe here, of
+  -- course: infinite lists.)
+  mem :: A.Array Int [MaxRoseTree m force]
+  mem = A.listArray (0, maxIdx)
+    [ [cutLoops <*> fmap (IM.mapWithKey memoizedNode) . lookupEdges $ (i, f) | f <- [toEnum 0..]]
+    | i <- [0..maxIdx]
+    ]
+
+  memoizedNode :: Int -> Max force -> MaxRoseNode m force
+  memoizedNode i (Max force) = MaxRoseNode force $ mem A.! i !! fromEnum force
+
+  -- And this is the function that prevents the search from actually being
+  -- infinite. It applies a filter to a `MaxRoseTree` at every level, looking for
+  -- indices anywhere in the tree that match the current vertex. If a match is
+  -- found with greater force than the current force, that part of the tree is
+  -- rewritten to abort; otherwise, that part of the tree is rewritten to be
+  -- empty (there's nothing new in that part of the search).
+  --
+  -- A new version of `cutLoops` is applied for each node in the search, so
+  -- each edge in a search path will add another filter on a new index. Since
+  -- there are a finite number of indices in our universe, this guarantees that
+  -- the analysis terminates, because no single search path can have length
+  -- greater than maxIdx.
+  cutLoops :: (Int, force) -> MaxRoseTree m force -> MaxRoseTree m force
+  cutLoops (i, force) = go
+    where
+    go = (=<<) . IM.traverseWithKey $ \i' (MaxRoseNode force' inner) ->
+      MaxRoseNode force' <$> if i == i' then guard (force >= force') $> pure IM.empty else pure $ go inner
```

Comment:
You got the internals correct; not sure about your first comment block. Remember that `MaxRoseTree m force` is a type alias for `m (MonoidalIntMap (MaxRoseNode m force))`; the outer `(=<<)` manages that `m`, since `traverseWithKey` gives us a Kleisli arrow (simplifying the types) `Map -> m Map`, and we want `go` to be a function `m Map -> m Map` (= `Tree -> Tree`).